### PR TITLE
Add foreman_puppet_autosign module

### LINKED ISSUE #766: Missing module for Puppet CA autosign entries
##### SUMMARY

There is no module that would allow the user to manage Puppet CA autosign entries.
##### ISSUE TYPE
 - Feature Idea

### PR DESCRIPTION
**Pull request by ogajduse**
Should fix #766.
Creating just a draft as it does not work yet.
I'd like to ask someone for a quick test of this PR.

When running the tests, for the first task I am getting the following:
```
File "/tmp/ansible_foreman_puppet_autosign_payload_699hdcad/ansible_foreman_puppet_autosign_payload.zip/ansible/modules/foreman_puppet_autosign.py", line 104, in <module>
  File "/tmp/ansible_foreman_puppet_autosign_payload_699hdcad/ansible_foreman_puppet_autosign_payload.zip/ansible/modules/foreman_puppet_autosign.py", line 100, in main
  File "/tmp/ansible_foreman_puppet_autosign_payload_699hdcad/ansible_foreman_puppet_autosign_payload.zip/ansible/module_utils/foreman_helper.py", line 922, in run
  File "/tmp/ansible_foreman_puppet_autosign_payload_699hdcad/ansible_foreman_puppet_autosign_payload.zip/ansible/module_utils/foreman_helper.py", line 474, in scope_for
TypeError: 'NoneType' object is not subscriptable
```

I was playing with it for some time and then I gave up. :worried: Is there any way, how to feed the `smart_proxy` parameter into `foreman_params`? Is there any other similar module, that I can inspire by?

Any help would be appreciated.

### CODE REVIEW

File: plugins/modules/foreman_puppet_autosign.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+class ForemanPuppetAutosignModule(ForemanEntityAnsibleModule):
+    pass
+
+
+def main():
+    module = ForemanPuppetAutosignModule(
+        foreman_spec=dict(
+            id=dict(type='str', required=True),
+            puppet_proxy=dict(type='entity', flat_name='smart_proxy', resource_type='smart_proxies', required=True),
```
Comment (mdellweg) [inline plugins/modules/foreman_puppet_autosign.py]: I might have an idea:

```suggestion
            smart_proxy=dict(type='entity', required=True, aliases=['puppet_proxy']),
```
The names kind of all need to be the entity type for this scope thing to work.

**Comment by ogajduse** [comment]
I tried with the patch you suggested.
```
The full traceback is:
  File "/tmp/ansible_foreman_puppet_autosign_payload_yphewxec/ansible_foreman_puppet_autosign_payload.zip/ansible/module_utils/foreman_helper.py", line 74, in inner
    return f(self, *args, **kwargs)
  File "/tmp/ansible_foreman_puppet_autosign_payload_yphewxec/ansible_foreman_puppet_autosign_payload.zip/ansible/module_utils/foreman_helper.py", line 385, in list_resource
    return self._resource_call(resource, 'index', params)['results']
```
In `list_resource` function, before calling `_resource_prepare_params`:
https://github.com/theforeman/foreman-ansible-modules/blob/08e8375b4dc45a0b39b60cea7d4ea1264170bc37/plugins/module_utils/foreman_helper.py#L422

```
>>> params
{'smart_proxy_id': 1, 'thin': False, 'search': 'name="None"', 'per_page': 4294967296}
>>> resource
'autosign'
```
After calling the function:
```
>>> params
{}
```
The rest of the traceback...
```
  File "/tmp/ansible_foreman_puppet_autosign_payload_yphewxec/ansible_foreman_puppet_autosign_payload.zip/ansible/module_utils/foreman_helper.py", line 352, in _resource_call
    return self._resource(resource).call(*args, **kwargs)
  File "/home/ogajduse/.virtualenvs/fam/lib/python3.7/site-packages/apypie/resource.py", line 30, in call
    return self.api.call(self.name, action, params, headers, options, data, files)
  File "/home/ogajduse/.virtualenvs/fam/lib/python3.7/site-packages/apypie/api.py", line 189, in call
    return self._call_action(action, params, headers, data, files)
  File "/home/ogajduse/.virtualenvs/fam/lib/python3.7/site-packages/apypie/api.py", line 196, in _call_action
    route.path_with_params(params),
  File "/home/ogajduse/.virtualenvs/fam/lib/python3.7/site-packages/apypie/route.py", line 27, in path_with_params
    raise KeyError("missing param '{}' in parameters".format(param))
```
So basically `_resource_prepare_params` is deleting the params. I see that it is trying to search for `'name="None"'`. This search is filled here.
https://github.com/theforeman/foreman-ansible-modules/blob/08e8375b4dc45a0b39b60cea7d4ea1264170bc37/plugins/module_utils/foreman_helper.py#L560

Any other suggestions? Am I missing something?

**Comment by mdellweg** [comment]
OK, smart_proxy lookup seems to work :ballot_box_with_check: 

Have a look at the apidoc `cat foreman.json | jq '.docs.resources.autosign.methods[1]'`.
The entity key should be `id` for `autosign`.
Try `entity_key` _and_ `entity_opts['search_by']` <- there is some confusion around those two atm.

**Comment by mdellweg** [comment]
OK, this `autosign` is _unique_ in a _new_ way.
It looks like you cannot search it, because its 'name' is `id`.
So it probably boils down to a `module.show_resource` with `failsafe=True` if it supports that.
Then you can skip the attempt to look it up again by something like `module.set_entity('entity', entity)`.

**Comment by mdellweg** [comment]
See `foreman_compute_attribute`, `foreman_external_usergroup`, `foreman_image`, `foreman_operatingsystem`, `foreman_os_default_template` and `foreman_smart_class_parameter` for an inspiration of special entity lookup hacks.

**Comment by evgeni** [comment]
Matthias' comments look spot on, so I'll leave this in "waiting on contributor" for now.

@ogajduse feel free to ping here or IRC if you need any guidance :)

**Comment by ogajduse** [comment]
I can not find time to finish this PR. Closing and leaving the branch as it is so anyone else can continue where I left off.

**Pull request closed without merging**